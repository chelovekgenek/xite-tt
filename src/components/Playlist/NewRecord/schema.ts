import * as Yup from "yup"

export const schema = Yup.object().shape({
  title: Yup.string()
    .min(2, "Too Short!")
    .max(32, "Too Long!")
    .required("Required"),
  artist: Yup.string()
    .min(2, "Too Short!")
    .max(32, "Too Long!")
    .required("Required"),
  src: Yup.string()
    .url("Not a url")
    .required("Required"),
})

export const initialValues: Yup.InferType<typeof schema> = {
  title: "",
  artist: "",
  src: "",
}
