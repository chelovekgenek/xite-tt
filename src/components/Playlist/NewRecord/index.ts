export { default as NewRecord } from "./NewRecord"
