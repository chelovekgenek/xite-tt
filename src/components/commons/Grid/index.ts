import { Row, Col } from "antd"

export const Grid = {
  Row,
  Col,
}
