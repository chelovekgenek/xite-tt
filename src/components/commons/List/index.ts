import { List as CList } from "./List"
import { Item as CItem } from "./Item"

export const List = Object.assign(CList, {
  Item: CItem,
})
