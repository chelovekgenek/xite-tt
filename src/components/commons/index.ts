export * from "./Modal"
export * from "./Icon"
export * from "./Grid"
export * from "./List"
