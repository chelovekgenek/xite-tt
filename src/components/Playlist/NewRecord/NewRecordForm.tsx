import React from "react"
import { Form } from "formik"

import { Input } from "components/commons/form"

export const NewRecordForm = () => (
  <Form>
    <Input label="Title" name="title" />
    <Input label="Artist" name="artist" />
    <Input label="Video URL" name="src" />
  </Form>
)
