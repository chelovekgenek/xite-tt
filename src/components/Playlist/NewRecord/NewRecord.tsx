import React, { useState, useCallback } from "react"
import { connect } from "react-redux"
import { withFormik, FormikProps } from "formik"

import { Icon, Modal, List } from "components/commons"
import { TAppState } from "store/entities"
import { VideoListActions } from "store/entities/video"

import { initialValues, schema } from "./schema"
import { NewRecordForm } from "./NewRecordForm"

interface IDispatchProps {
  append: typeof VideoListActions.append
}
interface IProps extends IDispatchProps {}

export const NewRecord: React.FC<IProps & FormikProps<typeof initialValues>> = ({
  isValid,
  handleSubmit,
  resetForm,
}) => {
  const [opened, setOpened] = useState(false)
  const toggleOpened = useCallback(() => setOpened(!opened), [opened])
  const handleOk = useCallback(() => {
    handleSubmit()
    toggleOpened()
  }, [handleSubmit, toggleOpened])
  const handleCancel = useCallback(() => {
    resetForm()
    toggleOpened()
  }, [resetForm, toggleOpened])
  return (
    <React.Fragment>
      <List.Item onClick={toggleOpened}>
        <Icon type="plus" />
        Add video
      </List.Item>
      <Modal
        okText="Submit"
        closable={false}
        visible={opened}
        onOk={handleOk}
        onCancel={handleCancel}
        okButtonProps={{ disabled: !isValid }}
      >
        <NewRecordForm />
      </Modal>
    </React.Fragment>
  )
}

export default connect<{}, IDispatchProps, {}, TAppState>(
  null,
  {
    append: VideoListActions.append,
  },
)(
  withFormik<IProps, typeof initialValues>({
    handleSubmit: (values, { props }) => props.append(values),
    validationSchema: schema,
    validateOnBlur: true,
  })(NewRecord),
)
