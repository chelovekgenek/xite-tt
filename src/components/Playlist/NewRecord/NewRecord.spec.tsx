import React from "react"
import { shallow } from "enzyme"

import { List, Modal } from "components/commons"
import formikMock from "mocks/formik-props"
import { NewRecord } from "./NewRecord"

describe("NewRecord", () => {
  const defaultValues = { title: "", artist: "", src: "" }
  const render = (props = {}) =>
    shallow(
      <NewRecord
        {...formikMock}
        append={jest.fn() as any}
        initialValues={defaultValues}
        values={defaultValues}
        {...props}
      />,
    )
  it("renders correctly", () => {
    const comp = render()
    expect(comp).toMatchSnapshot()
  })
  it('should show "submit" button as disabled, while form is not valid', () => {
    const comp = render()
    expect(comp.find(Modal).prop("okButtonProps")).toEqual({ disabled: true })
  })
  it('should set modal to visible, when "add record" button was clicked', () => {
    const comp = render()
    comp.find(List.Item).simulate("click")
    expect(comp.find(Modal).prop("visible")).toEqual(true)
  })
})
