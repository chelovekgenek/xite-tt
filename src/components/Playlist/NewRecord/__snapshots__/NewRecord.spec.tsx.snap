// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewRecord renders correctly 1`] = `
<Fragment>
  <styled.li
    onClick={[Function]}
  >
    <Icon
      type="plus"
    />
    Add video
  </styled.li>
  <Modal
    closable={false}
    confirmLoading={false}
    maskTransitionName="fade"
    okButtonProps={
      Object {
        "disabled": true,
      }
    }
    okText="Submit"
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    transitionName="zoom"
    visible={false}
    width={520}
  >
    <NewRecordForm />
  </Modal>
</Fragment>
`;
